### PR TITLE
Support Associative Array's Basic Constrained Randomization (#5658)

### PR DESCRIPTION
**Pull request by YilouWang**
This patch addresses two main points:

1. Refactoring the array class structure in verilated_random.h:
The previous implementation had separate classes for different types of arrays, but the `datap`, `emitType`, and similar functions were highly similar across these classes. This patch replaces them with a single `arrayTemplate `class to handle these common functionalities. Any type-specific differences are now encapsulated in the `record_arr` function. This change reduces code duplication, improves clarity, and makes the codebase easier to maintain.

2. Adding and enabling support for constrained randomization of associative arrays:
After this patch, Verilator can support basic constrained randomization for associative arrays. Here's what "basic" entails: 
-   Integral indices: 
Types like `int, longint, shortint, byte, bit`, and `logic` are now supported. However, for indices larger than 64 bits, Verilator uses `VlWide()`, which is currently not supported. Thus, indices must remain within 64 bits for now. 
-   String indices: 
String indices are now supported by converting strings to their ASCII representation. During the verilate phase, the string-to-ASCII numbers are masked to ensure they fit within 64 bits. This approach may lead to potential conflicts (e.g., two masked strings mapping to the same number), but I believe this is unavoidable. To mitigate this, a dictionary is maintained during the simulate phase to detect conflicts. If a conflict occurs, Verilator will issue a warning advising the user to choose different strings.
-   Class and user-defined type indices: 
According to the IEEE standard, `class `and` user-defined types` can also serve as indices. However, during testing, I encountered issues with how Verilator itself processes these indices. These issues need to be resolved in Verilator before constrained randomization for these cases can be supported.

Although this patch focuses on basic support, I believe it already covers the majority of user cases. Therefore, I am submitting this PR to update the code and provide this initial functionality.

### CODE REVIEW

File: src/V3Randomize.cpp [hunk 1]
```diff
@@ -724,6 +724,45 @@ class ConstraintExprVisitor final : public VNVisitor {
...
+            } else if (actual_width <= 16) {
+                fmt = "#x%4x";
+            } else if (actual_width <= 32) {
+                fmt = "#x%8x";
+            } else if (actual_width <= 64) {
+                fmt = "#x%16x";
+            } else {
+                nodep->v3warn(
+                    CONSTRAINTIGN,
+                    "Unsupported: Verilator does not currently support associative array index "
```
Comment (wsnyder) [inline src/V3Randomize.cpp:756]:
```suggestion
                    "Unsupported: Associative array index "
```

And we need a test to fire this warning, along with a test that turns it off (to make sure your code here works with it disabled, which I suspect it doesn't.)

File: include/verilated_random.h [hunk 2]
```diff
@@ -38,12 +39,15 @@ class ArrayInfo final {
     void* const m_datap;  // Reference to the array variable data
     const int m_index;  // Flattened (1D) index of the array element
     const std::vector<size_t> m_indices;  // Multi-dimensional indices of the array element
+    const std::vector<size_t> m_idx_widths;  // Multi-dimensional indices' bit widths
```
Comment (wsnyder) [inline include/verilated_random.h:42]:
```suggestion
    const std::vector<size_t> m_idxWidths;  // Multi-dimensional indices' bit widths
```

File: include/verilated_random.h [hunk 3]
```diff
@@ -90,25 +94,31 @@ class VlRandomVar VL_NOT_FINAL {
...
     void* datap(int idx) const override {
         const std::string indexed_name = name() + std::to_string(idx);
         const auto it = m_arrVarsRefp->find(indexed_name);
-        if (it != m_arrVarsRefp->end()) return it->second->m_datap;
-        return &static_cast<T*>(VlRandomVar::datap(idx))->atWrite(idx);
+        if (it != m_arrVarsRefp->end()) {
+            return it->second->m_datap;
+        } else {
+            VL_FATAL_MT(__FILE__, __LINE__, "randomize",
+                        "Error: indexed_name not found in m_arr_vars");
```
Comment (wsnyder) [inline include/verilated_random.h:110]:
```suggestion
                        "indexed_name not found in m_arr_vars");
```

File: include/verilated_random.h [hunk 4]
```diff
@@ -90,25 +94,31 @@ class VlRandomVar VL_NOT_FINAL {
...
     }
-    void emitSelect(std::ostream& s, const std::vector<size_t>& indices) const {
+    void emitSelect(std::ostream& s, const std::vector<size_t>& indices,
+                    const std::vector<size_t>& idx_widths) const {
         for (size_t idx = 0; idx < indices.size(); ++idx) s << "(select ";
         s << name();
         for (size_t idx = 0; idx < indices.size(); ++idx) {
             s << " #x";
-            for (int j = 28; j >= 0; j -= 4) {
+            size_t bit_width = idx_widths[idx];
```
Comment (wsnyder) [inline include/verilated_random.h:120]:
```suggestion
            const size_t bit_width = idx_widths[idx];
```

File: include/verilated_random.h [hunk 5]
```diff
@@ -121,74 +131,29 @@ class VlRandomQueueVar final : public VlRandomVar {
             const auto it = m_arrVarsRefp->find(indexed_name);
             if (it != m_arrVarsRefp->end()) {
                 const std::vector<size_t>& indices = it->second->m_indices;
-                emitSelect(s, indices);
+                const std::vector<size_t>& idx_widths = it->second->m_idx_widths;
+                emitSelect(s, indices, idx_widths);
+            } else {
+                VL_FATAL_MT(__FILE__, __LINE__, "randomize",
+                            "Error: indexed_name not found in m_arr_vars");
```
Comment (wsnyder) [inline include/verilated_random.h:138]:
```suggestion
                            "indexed_name not found in m_arr_vars");
```

Similar other places (FATAL code prints %Error itself).

File: include/verilated_random.h [hunk 6]
```diff
@@ -231,6 +201,43 @@ class VlRandomizer final {
...
+    typename std::enable_if<!std::is_integral<T_Key>::value
+                            && !std::is_same<T_Key, std::string>::value>::type
+    process_key(const T_Key& key, std::string& indexed_name, size_t& integral_index,
+                const std::string& base_name, size_t& idx_width) {
+        VL_FATAL_MT(__FILE__, __LINE__, "randomize",
+                    "Unsupported: Only integral and string index of associative array is "
+                    "supported currently.");
+    }
+    size_t string_to_integral(const std::string& str) {
+        size_t result = 0;
```
Comment (wsnyder) [inline include/verilated_random.h:230]:
```suggestion
        uint64_t result = 0;
```

File: include/verilated_random.h [hunk 7]
```diff
@@ -231,6 +201,43 @@ class VlRandomizer final {
...
+        for (char c : str) {
+            result = (result << 8) | static_cast<size_t>(c);
+            result &= 0xFFFFFFFFFFFFFFFF;
```
Comment (wsnyder) [inline include/verilated_random.h:233]:
```suggestion
            result &= 0xFFFFFFFFFFFFFFFFULL;
```

But, I don't understand why this is needed, it seems a nop given it's a 64 bit variable.

File: include/verilated_random.h [hunk 8]
```diff
@@ -231,6 +201,43 @@ class VlRandomizer final {
...
+        }
+        if (seen_values.count(result) > 0 && seen_values[result] != str)
+            VL_WARN_MT(__FILE__, __LINE__, "randomize",
+                       "Conflict detected: Different strings mapped to the same 64-bit index.");
+        seen_values[result] = str;
```
Comment (wsnyder) [inline include/verilated_random.h:238]: Will this get slow at runtime?

**Comment by wsnyder** [comment]
I think having it in VL_DEBUG is ok, I'm hoping we'll soon relax the 64 bit restriction.

**Comment by YilouWang** [comment]
Thank you for the feedback! I'll start working on relaxing the 64-bit restriction soon. :)

Should I open a new issue for this, including the challenges with class and user-defined data types as indices, or we can still leave #5658 unsolved ?

**Comment by wsnyder** [comment]
>Should I open a new issue for this, including the challenges with class and user-defined data types as indices, or we can still leave https://github.com/verilator/verilator/issues/5658 unsolved ?

If you'd like to get #5658 merged once ready, which seems reasonable & gets it in, please make a new issue.  If you prefer to make #5658 one more complex before merging to resolve the user-defined types etc, you don't need a new issue.

**Comment by YilouWang** [comment]
Okay, Thanks. I will open a new issue then.

**Comment by YilouWang** [comment]
Hi, @wsnyder 

Would it be appropriate to push the code now, or are there any areas that still need improvement?

**Pull request merged**